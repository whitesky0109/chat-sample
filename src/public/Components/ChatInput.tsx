import React, { Component, ChangeEvent, KeyboardEvent } from 'react';

// Models
import { IMessage } from 'models/client';

export interface Props {
  onSendMessage: (msg: IMessage) => void;
}

export interface States {
  value: string;
}

export default class ChatInput extends Component<Props, States> {
  state: States = {
    value: '',
  };

  sendMessage(type: string, message: string | File) {
    const { onSendMessage } = this.props;

    if (onSendMessage) {
      onSendMessage({
        type,
        message,
      });
    }
    this.setState({ value: '' });
  }

  inputChange = (e: ChangeEvent<HTMLInputElement>) => {
    const { value } = e.target;

    this.setState({ value });
  }

  inputKeyPress = (e:KeyboardEvent<HTMLInputElement>) => {
    if (e.keyCode === 13 && !e.shiftKey) {
      e.preventDefault();
      this.onSend();
    }
  }

  fileChange = (e: ChangeEvent<HTMLInputElement>) => {
    const { files } = e.target;

    if (files) {
      for (let i = 0; i < files.length; i += 1) {
        const file = files.item(i);

        if (file) {
          this.sendMessage(file.type, file);
        }
      }
    }
  }

  onSend = () => {
    const { value } = this.state;

    if (value) {
      this.sendMessage('text', value);
    }
  }

  renderFileInput() {
    return <>
    <div className="file-input" >
      <label htmlFor="image" >
          <span className="fa fa-upload" aria-hidden="true" />
      </label>
      <input type="file" accept="image/*" id="image" className="image-input" onChange={this.fileChange} />
    </div>
    </>;
  }

  render() {
    const { value } = this.state;

    return <div className="chat-input clearfix">
      {this.renderFileInput()}

      <input
        type="text"
        value={value}
        onChange={this.inputChange}
        onKeyDown={this.inputKeyPress}
        placeholder="Type your message"
      />

      <button onClick={this.onSend}>Send</button>
    </div>;
  }
}
