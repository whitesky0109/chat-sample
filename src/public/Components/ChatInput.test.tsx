import React from 'react';
import { mount } from 'enzyme';
import ChatInput from './ChatInput';

describe('ChatInput', () => {
  let component: any = null;
  const mockOnSendMessage = jest.fn();

  it('renders correctly', () => {
    component = mount(<ChatInput onSendMessage={mockOnSendMessage} />);
  });

  it('텍스트 입력 변경 확인', () => {
    const input = component.find('input[type="text"]');
    const inputValue = 'abcdefg';

    expect(input).not.toBeUndefined();
    input.simulate('change', { target: { value: inputValue } });

    expect(component.state().value).toBe(inputValue);
  });

  it('입력 전달 확인', () => {
    const input = component.find('input[type="text"]');
    input.simulate('keydown', { keyCode: 13 });
    expect(mockOnSendMessage.mock.calls.length).toBe(1);
  });

  it('입력 자동 비움 확인', () => {
    expect(component.state().value).toBe('');
  });

  it('미입력 미전달 확인', () => {
    const input = component.find('input[type="text"]');
    input.simulate('keydown', { keyCode: 13 });
    expect(mockOnSendMessage.mock.calls.length).toBe(1);
  });

  it('이미지 전달 확인', () => {
    const file = new File(['(⌐□_□)'], 'chucknorris.png', { type: 'image/png' });

    component.find('input[type="file"]').simulate('change', { target: { files: [file] } });
    expect(mockOnSendMessage.mock.calls.length).toBe(2);
  });

  it('matches snapshot', () => {
    expect(component).toMatchSnapshot();
  });

});
