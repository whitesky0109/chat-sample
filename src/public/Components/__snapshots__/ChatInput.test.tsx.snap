// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChatInput matches snapshot 1`] = `
<ChatInput
  onSendMessage={
    [MockFunction] {
      "calls": Array [
        Array [
          Object {
            "message": "abcdefg",
            "type": "text",
          },
        ],
        Array [
          Object {
            "message": File {},
            "type": "image/png",
          },
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
>
  <div
    className="chat-input clearfix"
  >
    <div
      className="file-input"
    >
      <label
        htmlFor="image"
      >
        <span
          aria-hidden="true"
          className="fa fa-upload"
        />
      </label>
      <input
        accept="image/*"
        className="image-input"
        id="image"
        onChange={[Function]}
        type="file"
      />
    </div>
    <input
      onChange={[Function]}
      onKeyDown={[Function]}
      placeholder="Type your message"
      type="text"
      value=""
    />
    <button
      onClick={[Function]}
    >
      Send
    </button>
  </div>
</ChatInput>
`;
